mint-auditor/src/gnosis/api_data_types.rs

```diff
@@ -0,0 +1,201 @@
+    pub tx_hash: EthTxHash,
+
+    /// Decoded transaction data
+    #[serde(rename = "dataDecoded")]
+    pub data_decoded: Option<DataDecoded>,
+}
+
+/// Ethereum transfer
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct EthereumTransfer {
+    /// From address
+    pub from: EthAddr,
+
+    /// To address
+    pub to: EthAddr,
+
+    /// Token contract address that is being transferred
+    /// None for Eth transfers
+    #[serde(rename = "tokenAddress")]
+    pub token_address: Option<EthAddr>,
+
+    /// Transaction hash
+    #[serde(rename = "transactionHash")]
+    pub tx_hash: EthTxHash,
+
+    /// Transaction type
+    #[serde(rename = "type")]
+    pub tx_type: String,
+
+    /// Value being transferred
+    pub value: JsonU64,
+}
+
+/// Ethereum transaction
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct EthereumTransaction {
+    /// Transaction hash
+    #[serde(rename = "txHash")]
+    pub tx_hash: EthTxHash,
+
+    /// Ethereum block number.
+    #[serde(rename = "blockNumber")]
+    pub eth_block_number: u64,
+
+    /// Transfers
+    pub transfers: Vec<EthereumTransfer>,
+}
+
+/// Possible transaction types that are returned from
+/// `/safes/{address}/all-transactions/`
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(tag = "txType")]
+pub enum Transaction {
+    /// Multi-sig transaction
+    #[serde(rename = "MULTISIG_TRANSACTION")]
+    MultiSig(MultiSigTransaction),
+
+    /// Ethereum transaction
+    #[serde(rename = "ETHEREUM_TRANSACTION")]
+    Ethereum(EthereumTransaction),
+
+    /// Module transaction
+    #[serde(rename = "MODULE_TRANSACTION")]
+    Module(Value),
+}
+
+/// Raw (unparsed) transaction.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RawGnosisTransaction {
+    raw: Value,
+}
+
+impl RawGnosisTransaction {
+    /// Decode a Gnosis Safe transaction.
+    pub fn decode(&self) -> Result<Transaction, Error> {
+        Ok(serde_json::from_value(self.raw.clone())?)
+    }
+
+    /// Get the transaction hash.
+    pub fn tx_hash(&self) -> Result<EthTxHash, Error> {
+        let hash_str = self
+            .raw
+            .get("transactionHash")
+            .or_else(|| self.raw.get("txHash"))
+            .and_then(|val| val.as_str())
+            .ok_or_else(|| {
+                Error::Other("GnosisSafeTransaction: missing transactionHash".to_string())
+            })?;
+        EthTxHash::from_str(hash_str)
+            .map_err(|err| Error::Other(format!("Failed parsing tx hash '{}': {}", hash_str, err)))
+    }
+
+    /// Serialize transaction into JSON.
+    pub fn to_json_string(&self) -> String {
+        self.raw.to_string()
+    }
+
+    /// Deserialize transaction from JSON.
+    pub fn from_json_bytes(bytes: &[u8]) -> Result<Self, Error> {
+        Ok(Self::from(serde_json::from_slice::<Value>(bytes)?))
+    }
+}
+
+impl From<Value> for RawGnosisTransaction {
+    fn from(value: Value) -> Self {
+        Self { raw: value }
```

Comment:
Optional/DRY: Name the parameter like the field.

```suggestion
    fn from(raw: Value) -> Self {
        Self { raw }
```